### PR TITLE
PROTOTECH-57: Collateral Can Be Extracted Without Redeeming LP

### PR DESCRIPTION
**Pull request by grandizzy**
# Description of change
## High level
* additional protection for https://github.com/Fixed-Point-Solutions/prototech-ajna-audit/issues/57
  * revert when LP to amount or amount to LP result is 0:
    * `addQuoteToken`: revert with `InvalidAmount()` if calculated `unscaledAmount` is zero
    * `removeCollateral`: revert with `InsufficientLP()` if calculated `LP` amount is zero
    * `mergeOrRemoveCollateral`: (used by `ERC721Pool.mergeOrRemoveCollateral`) revert with `InvalidAmount()` if calculated amount of collateral to remove is 0
    * `_removeMaxCollateral`: (used by `ERC20Pool.removeCollateral`) revert with `InsufficientLP()` if calculated `LP` amount is 0
    * `_removeMaxDeposit`: (used by `ERC20Pool.moveQuoteToken` and `ERC20Pool.removeQuoteToken`) revert with `InsufficientLP()` if calculated `LP` amount is zero; revert with `InvalidAmount()` if calculated unscaled amount of quote token to remove is 0
    * `kickWithDeposit`: revert with `InsufficientLiquidity()` if calculated unscaled amount to use on kick is 0; revert with `InsufficientLP()` if LP redeemed amount to kick auction is 0. Extra fix: gas optimization by reusing local `vars.lenderLP` instead reloading from storage

# Description of bug or vulnerability and solution
* See https://github.com/Fixed-Point-Solutions/prototech-ajna-audit/issues/57
* Extra protection to complement rounding fixes - always ensure calculated amounts / redeemed LP are 0, else revert

# Contract size
## Pre Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,470B  (99.56%)
  ERC20Pool                -  23,819B  (96.92%)
```
## Post Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,470B  (99.56%)
  ERC20Pool                -  23,819B  (96.92%)
```

# Gas usage
## Pre Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120727          | 177942 | 162586 | 712465  | 60004   |
| bucketTake                           | 308615          | 326982 | 326982 | 345350  | 4       |
| drawDebt                             | 250455          | 295356 | 269279 | 809196  | 136003  |
| kick                                 | 147490          | 221511 | 220012 | 1032365 | 48000   |
| kickWithDeposit                      | 196696          | 274344 | 270489 | 996938  | 8000    |
| moveQuoteToken                       | 278302          | 278302 | 278302 | 278302  | 1       |
| removeQuoteToken                     | 144953          | 167579 | 171672 | 195094  | 4000    |
| repayDebt                            | 574996          | 607152 | 597044 | 755414  | 16002   |
| settle                               | 349355          | 349355 | 349355 | 349355  | 1       |
| take                                 | 81142           | 82544  | 82236  | 313980  | 15998   |
```
## Post Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120799          | 178014 | 162658 | 712537  | 60004   |
| bucketTake                           | 308615          | 326982 | 326982 | 345350  | 4       |
| drawDebt                             | 250455          | 295356 | 269279 | 809196  | 136003  |
| kick                                 | 147490          | 221511 | 220012 | 1032365 | 48000   |
| kickWithDeposit                      | 196696          | 274344 | 270489 | 996938  | 8000    |
| moveQuoteToken                       | 278348          | 278348 | 278348 | 278348  | 1       |
| removeQuoteToken                     | 144990          | 167621 | 171718 | 195140  | 4000    |
| settle                               | 349355          | 349355 | 349355 | 349355  | 1       |
| take                                 | 81142           | 82544  | 82236  | 313980  | 15998   |
```

### CODE REVIEW

File: src/libraries/external/KickerActions.sol [hunk 1]
```diff
@@ -219,11 +224,18 @@ library KickerActions {
...
+
+            // revert if calculated unscaled amount is 0
+            if (unscaledAmountToRemove == 0) revert InsufficientLiquidity();
+
             Deposits.unscaledRemove(deposits_, index_, unscaledAmountToRemove);
             vars.bucketUnscaledDeposit -= unscaledAmountToRemove;
         }
 
-        vars.redeemedLP = Maths.min(lender.lps, vars.redeemedLP);
+        vars.redeemedLP = Maths.min(vars.lenderLP, vars.redeemedLP);
```
Comment (grandizzy) [inline src/libraries/external/KickerActions.sol:235]: extra fix/ gas optimization: reused `vars.lenderLP`

**Review by EdNoepel — APPROVED** [review]
LGTM

**Review by prat-gpt — APPROVED** [review]
LGTM!

**Review by godsflaw — APPROVED** [review]
LGTM, this PR also resolved a number of regressions in https://github.com/ajna-finance/contracts/pull/810

File: src/libraries/external/LenderActions.sol [hunk 2]
```diff
@@ -580,6 +589,9 @@ library LenderActions {
                 fromIndex
             );
 
+            // revert if calculated amount of collateral to remove is 0
+            if (collateralRemoved == 0) revert InvalidAmount();
```
Comment (mattcushman) [inline src/libraries/external/LenderActions.sol:593]: Shouldn't we get the `lpRemoved_` return value from `_removeMaxCollateral(...)` above, and also revert if that `==0`?

Comment (mattcushman) [reply src/libraries/external/LenderActions.sol:593]: nvm... I think that `_removeMaxCollateral(...)` itself will revert in this case.

**Review by mattcushman — APPROVED** [review]
lgtm... in the process of reviewing i noted some other rounding improvements we can look at though, will follow up

**Comment by grandizzy** [comment]
> lgtm... in the process of reviewing i noted some other rounding improvements we can look at though, will follow up

@prateek105 added a similar comment here https://github.com/ajna-finance/contracts/pull/838#issuecomment-1559324851

File: src/libraries/external/LenderActions.sol [hunk 3]
```diff
@@ -184,7 +185,11 @@ library LenderActions {
         // revert if (due to rounding) the awarded LP is 0
         if (bucketLP_ == 0) revert InsufficientLP();
 
-        Deposits.unscaledAdd(deposits_, params_.index, Maths.wdiv(addedAmount, bucketScale));
+        uint256 unscaledAmount = Maths.wdiv(addedAmount, bucketScale);
+        // revert if unscaled amount is 0
+        if (unscaledAmount == 0) revert InvalidAmount();
```
Comment (kmbarry1) [inline src/libraries/external/LenderActions.sol:190]: So I think these checks are valuable precautions and should at least eliminate some fraction of potentially dangerous rounding errors. However, there may still exist cases where the rounding error is large enough to matter even if the amounts aren't exactly zero.

Stronger protections include (1) rounding against the user (which I think was added in most places in  https://github.com/ajna-finance/contracts/pull/838), and (2) explicitly checking the truncation error is not larger than some % bound.

Example for (2):
```
uint256 unscaledAmount = Maths.wdiv(addedAmount, bucketScale);
require(9990 * addedAmount / 10_000 <= unscaledAmount && 10_010 * addedAmount / 10_000)
```

I think (2) might introduce annoying edge cases where trivial amounts can't be manipulated, so not necessarily recommending it without more research.

**Pull request merged**